### PR TITLE
feat: load node configurations using gw_get_node_info

### PR DESCRIPTION
**Pull request by keroro520**
web3-indexer get chain spec via `gw_get_node_info` as well.

### CODE REVIEW

**Review by RetricSu — COMMENTED** [review]
please update the README as well https://github.com/nervosnetwork/godwoken-web3#config-indexer

**Comment by keroro520** [comment]
# Godwoken Web3 API

A Web3 RPC compatible layer build upon Godwoken/Polyjuice.

## Development

### Config database

```bash
$ cat > ./packages/api-server/.env <<EOF
DATABASE_URL=postgres://username:password@localhost:5432/your_db
REDIS_URL=redis://user:password@localhost:6379 <redis url, optional, default to localhost on port 6379>

GODWOKEN_JSON_RPC=<godwoken rpc>
GODWOKEN_READONLY_JSON_RPC=<optional, default equals to GODWOKEN_JSON_RPC>

SENTRY_DNS=<sentry dns, optional>
SENTRY_ENVIRONMENT=<sentry environment, optional, default to `development`>,
NEW_RELIC_LICENSE_KEY=<new relic license key, optional>
NEW_RELIC_APP_NAME=<new relic app name, optional, default to 'Godwoken Web3'>

PG_POOL_MAX=<pg pool max count, optional, default to 20>
CLUSTER_COUNT=<cluster count, optional, default to num of cpus>
GAS_PRICE_CACHE_SECONDS=<seconds, optional, default to 0, and 0 means no cache>
EXTRA_ESTIMATE_GAS=<eth_estimateGas will add this number to result, optional, default to 0>
ENABLE_CACHE_ETH_CALL=<optional, enable eth_call cache, default to false>
LOG_LEVEL=<optional, allowed value: `debug` / `info` / `warn` / `error`, default to `debug` in development, and default to `info` in production>
LOG_FORMAT=<optional, allowed value: `json`>
MAX_SOCKETS=<optional, max number of httpAgent sockets per host for web3 connecting to godwoken, default to 10>
WEB3_LOG_REQUEST_BODY=<optional, boolean, if true, will log request method / body, default to false>
PORT=<optional, the api-server running port, default to 8024>
EOF

$ yarn

# For api-server & indexer
$ DATABASE_URL=<your database url> make migrate

# Only for test purpose
$ yarn workspace @godwoken-web3/api-server reset_database
```

rate limit config

```bash
$ cat > ./packages/api-server/rate-limit-config.json <<EOF
{
  "expired_time_milsec": 60000,
  "methods": {
    "poly_executeRawL2Transaction": 30,
    "<rpc method name>": <max requests number in expired_time>
  }
}
EOF
```

### Start Indexer

```bash
cargo build --release

godwoken_rpc_url=<godwoken rpc, e.g. "http://godwoken:8119"> \
ws_rpc_url=<godwoken websocket rpc, e.g. "ws://godwoken:8120"> \
pg_url=<database url, e.g. "postgres://username:password@localhost:5432/dbname"> \
./target/release/gw-web3-indexer
```

### Start API server

```bash
yarn run build:godwoken
yarn run start
```

#### Start in production mode

```bash
yarn run build && yarn run start:prod
```

#### Start via pm2

```bash
yarn run build && yarn run start:pm2
```

#### Start using docker image

```bash
docker run -d -it -v <YOUR .env FILE PATH>:/godwoken-web3/packages/api-server/.env  -w /godwoken-web3  --name godwoken-web3 nervos/godwoken-web3-prebuilds:<TAG> bash -c "yarn workspace @godwoken-web3/api-server start:pm2"
```

then you can monit web3 via pm2 inside docker container:

```bash
docker exec -it <CONTAINER NAME> /bin/bash
```
```
$ root@ec562fe2172b:/godwoken-web3# pm2 monit
```

Http url: http://your-url/

WebSocket url: ws://your-url/ws

### Docker Prebuilds

local development:

```sh
make build-test-image # (tag: latest-test)
```

push to docker:

```sh
make build-push # needs login, will ask you for tag
```

resource:

- docker image: https://hub.docker.com/repository/docker/nervos/godwoken-web3-prebuilds
- code is located in `/godwoken-web3` with node_modules already installed and typescript compiled to js code.


### RPC Docs

Get RPC docs at [RPCs doc](docs/apis.md)


> please update the README as well https://github.com/nervosnetwork/godwoken-web3#config-indexer

Thanks. [0bb528d](https://github.com/nervosnetwork/godwoken-web3/pull/369/commits/0bb528d4cfe054d26f94a9165a3bda55ada53f5c) update the indexer's document. Please take a look.

**Review by Flouse — APPROVED** [review]

File: README.md [hunk 1]
```diff
@@ -54,33 +54,14 @@ $ cat > ./packages/api-server/rate-limit-config.json <<EOF
...
-EOF
-```
-
-Or just run script, generate configs from `packages/api-server/.env` file and godwoken rpc `gw_get_node_info`.
-
-```bash
-node scripts/generate-indexer-config.js <websocket rpc url>
-```
-
 ### Start Indexer
```
[the remark this answers is not in the document]
Comment (Flouse) [reply README.md:58]: I think sometimes we still need the configuration file to set other configs such as `sentry`.

File: crates/indexer/src/config.rs [hunk 2]
```diff
@@ -55,9 +58,80 @@ impl Display for IndexerConfig {
...
+            path.as_ref().to_string_lossy().to_string()
+        );
+    }
+
+    // Load components configurations from environment variables
+    let pg_url = env::var("pg_url")
+        .unwrap_or_else(|_| "postgres://postgres:postgres@127.0.0.1:5432/".to_string());
+    let godwoken_rpc_url =
+        env::var("godwoken_rpc_url").unwrap_or_else(|_| "http://127.0.0.1:8119".to_string());
+    let ws_rpc_url = env::var("ws_rpc_url").unwrap_or_else(|_| "http://127.0.0.1:8120".to_string());
```
[the remark this answers is not in the document]
Comment (keroro520) [reply crates/indexer/src/config.rs:82]: I want to keep compatible with previous.
So let's do it later?

File: crates/indexer/src/config.rs [hunk 3]
```diff
@@ -55,9 +58,80 @@ impl Display for IndexerConfig {
...
+        dotenv::from_path(path)?;
+    } else {
+        log::info!(
+            "Cannot find configuration file {}, continue",
+            path.as_ref().to_string_lossy().to_string()
+        );
+    }
+
+    // Load components configurations from environment variables
+    let pg_url = env::var("pg_url")
```
[the remark this answers is not in the document]
Comment (keroro520) [reply crates/indexer/src/config.rs:78]: Okay.

Remove default values for `godwoken_rpc_url` as well?

**Review by Flouse — APPROVED** [review]

File: crates/indexer/src/config.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (classicalliu) [reply crates/indexer/src/config.rs:78]: I think no need to remove default `godwoken_rpc_url` .

File: crates/indexer/src/config.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (classicalliu) [reply crates/indexer/src/config.rs:82]: Ohh, sorry, I see, it's not an `Option`, so keep this default value is ok.

File: crates/indexer/src/config.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (keroro520) [reply crates/indexer/src/config.rs]: d5c3d4f removes default value for `pg_url`.

**Review by RetricSu — APPROVED** [review]

**Review by classicalliu — APPROVED** [review]

**Pull request merged**